### DISCUSSION TITLE
Cart Block: registerCheckoutFilters | cartItemPrice | HTML

### DISCUSSION THREAD

**Discussion by larsenlarsson** [woocommerce/woocommerce#45004] [WooCommerce, Blocks, and Site Editing]
Dear WooCommerce team, we hope you are doing well and can help us.

In the previous cart, which is included via shortcode, we have used the hook "woocommerce_cart_item_subtotal" to get additional output regarding the tax. Simplified, we did it like this:

```
add_filter( 'woocommerce_cart_item_subtotal', function( $cart_item_subtotal, $cart_item, $cart_item_key ) {
	return $cart_item_subtotal . my_get_cart_item_tax_string( $cart_item );
}, 10, 3 );
```

By doing so, we achieved such an output, for example:
![Bildschirmfoto 2023-08-23 um 13 02 32](https://github.com/woocommerce/woocommerce-blocks/assets/2648926/ae10b6af-ece1-4635-87c6-99ff6fce7f87)

This filter no longer works in the cart block. We then followed this documentation to get the output: https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-filters.md

We now use "registerCheckoutFilters" to filter "cartItemPrice":

```
registerCheckoutFilters( 'myTaxOutput', {
	cartItemPrice: ( value, extensions, args ) => {

		let line_item_tax = '';
		let item_data = args.cartItem.item_data;
		
		if ( item_data ) {
			let line_item_tax_object = item_data.find( x => x.key === '_my_cart_item_subotal_tax_data' );
			if ( line_item_tax_object ) {
				line_item_tax = line_item_tax_object.value;
			}
		}

		if ( '' !== line_item_tax ) {
			value = value + ' ' + line_item_tax;
		}
		return value;
	},
	
} );
```

With the hook "woocommerce_get_item_data" we added the desired output to "item_data" via PHP:

```
add_action( 'woocommerce_get_item_data', function( $data, $cart_item ) {

	$data[] = array(
		'key' => '_my_cart_item_subotal_tax_data',
		'value' => my_get_cart_item_tax_string( $cart_item ),
		'hidden' => true
	);
	return $data;
}, 10, 2 );
```

Basically, the output in the cart block works as desired:
![Bildschirmfoto 2023-08-23 um 13 12 06](https://github.com/woocommerce/woocommerce-blocks/assets/2648926/7287a588-fddc-431d-ac4d-a2043a855444)

Our problem now is that any HTML that the "my_get_cart_item_tax_string" function actually outputs is removed.

If we add it manually in the JavaScript, it is "escaped":
```
registerCheckoutFilters( 'myTaxOutput', {
	cartItemPrice: ( value, extensions, args ) => {

		let line_item_tax = '';
		let item_data = args.cartItem.item_data;
		
		if ( item_data ) {
			let line_item_tax_object = item_data.find( x => x.key === '_my_cart_item_subotal_tax_data' );
			if ( line_item_tax_object ) {
				line_item_tax = line_item_tax_object.value;
			}
		}

		if ( '' !== line_item_tax ) {
			value = value + '<span>' + line_item_tax + '</span>';
		}
		return value;
	},
	
} );
```
![Bildschirmfoto 2023-08-23 um 13 14 47](https://github.com/woocommerce/woocommerce-blocks/assets/2648926/9e8f30ac-a99d-428d-9bef-df3bac0fa5ca)

Do you have any hints or other help for us on how to achieve the output with HTML here? Unfortunately our REACT and block knowledge is still "under construction" ;)

Thanks in advance for any help.

**Comment by senadir** [comment]
We sadly do not support HTML based entries now for filters due to stability reasons so for now, you will have to continue using it as is sadly.

**Comment by larsenlarsson** [comment]
Hi @senadir, 

thank you very much for your time and your answer. 
That's a great pity, the store admins will surely wish to be able to style our additional output as before.
Do you see any possibility that this will work in the future?
Thanks in advance for your efforts.

**Comment by LuigiPulcini** [comment]
We have several plugins that interact with the cart by customizing the content of the line item, via the `woocommerce_get_item_data` filter. As much as we would love to embrace the new block design, our only option so far was to recommend our customers move away from the cart/checkout blocks and restore the traditional shortcodes on those pages, which is truly a shame.

I was following [this thread](https://github.com/woocommerce/woocommerce-blocks/issues/10435) and @senadir [wrote](https://github.com/woocommerce/woocommerce-blocks/issues/10435#issuecomment-1688575795):

> HTML is stripped as a measure of security and to preserve the layout/style of the cart. As meta data is only supposed to be a key: value of strings.

Please allow me to break down the rationale behind this choice:

1. **a measure of security**: I would like to have more details about this reason and why we cannot find solutions similar to what `wp_kses()` does on the server side (besides, the content we are printing comes from the server side anyway, because it runs through the `woocommerce_get_item_data`);
2. **preserve the layout/style of the cart**: prohibiting HTML content seems too drastic when it should be (and has always been) up to the developer and even in their own interest to ensure the styling is preserved and any additional element is well harmonized with the surrounding environment;
3. **a key: value of strings**: metadata are strings for sure **BUT** the way they are presented should be left in the hands of us developers/designers. For example, no one argues that a URL is a string but I might want to present it as a clickable link, which is not possible in the current implementation of the cart/checkout blocks (see the example below).

![image](https://github.com/user-attachments/assets/f539ce27-b16d-4dd9-8adf-48bc2d14e2c3)

One of our plugins adds several option fields to the add-to-cart form of a product. Among them is a File upload option. When adding a fully-configured product to the cart, the plugin uses the metadata to recap the configuration, including a link to the uploaded image so customers can review each item and ensure everything is correct. That works great in the cart/checkout shortcodes but becomes useless plain text in the blocks.

And what if I wanted to present that metadata as a clickable thumbnail of the uploaded image, which would be even more effective for visually confirming the purchase?
* Very easy to accomplish in the traditional cart/checkout shortcodes
* Just impossible in the new cart/checkout blocks.

I wonder how many developers are facing similar issues and are forced to suggest their customers to ditch the cart/checkout blocks and restore the traditional shortcodes.

**Comment by LuigiPulcini** [comment]
Is anyone from the Woocommerce team available at least to acknowledge the issues discussed above and give us directions on the best approach? Unlike other experimental areas that are still under development, Cart and Checkout block-based pages are now the default experience for new installations and every plugin customizing the content of the cart item metadata with HTML is going to break with those blocks.

**Comment by ralucaStan** [comment]
In https://github.com/woocommerce/woocommerce/pull/61437 we've allowed some HTML elements in the `item_data` metadata: `'a', 'b', 'em', 'i', 'strong', 'br', 'abbr', 'span'`.

For now, we're keeping this as a list of allowed elements. 
If you were using `item_data` to add other HTML elements please share your use case, if you haven't already.
